feat: add mechanics to provide classes and factories with internally registered tokens

yurahuchok opened the pull request:
## 🏷️ Internally registered tokens / Known Classes and Known Factories

You can register class tokens internally in the classes that need to be provided, and then provide such classes by using the `provideClass` method's overload. To implement the "Known Class", add a `static knownAs` property with a string literal value. This way, it is possible to create a hierarchy of class dependencies without relying on arbitrarily specified tokens, instead opting to consolidate said tokens within the classes themselves.

```ts
import { createInjector } from 'typed-inject';

class Foo {
  static knownAs = "FooToken" as const;
}

class Bar {
  static knownAs = "BarToken" as const;

  static inject = [ Foo.knownAs ] as const;

  constructor(foo: Foo) {}
}

const injector = createInjector()
  .provideClass(Foo)
  .provideClass(Bar);
```

Similarly, it works with `provideFactory` method's overload.

```ts
import { createInjector } from 'typed-inject';

function fooFactory() {
  return "foo factory called";
}
fooFactory.knownAs = 'Foo' as const;

function barFactory() {
  return "bar factory called";
}
barFactory.inject = [ fooFactory.knownAs ] as const;
barFactory.knownAs = "Bar" as const;

const injector = createInjector()
  .provideFactory(fooFactory)
  .provideFactory(barFactory);
```

yurahuchok committed:
feat: add mechanics to provide classes with internally specified tokens

yurahuchok committed:
Update README.md

yurahuchok committed:
Update README.md

yurahuchok committed:
Update README.md

nicojs committed:
feat(node): drop support for node 16 (#71)

BREAKING CHANGE: Please use Node 18 or higher

mayorandrew committed:
feat: improve error messages (#66)

- Simplify error messages
- Allow for near-infinite deep dependency trees.

Co-authored-by: Nico Jansen <jansennico@gmail.com>

nicojs commented:
This is very interesting. I've got a couple of remarks.

1. Could we also implement this for `provideFactory`?
1. Could we use overloading on the existing `provideClass` and `provideFactory` functions? I find the "provideInjectableXxx" naming scheme confusing, as classes provided by `provideClass` are also injectable.
1. Could we rename `injectableAs` to `knownAs`? Or maybe `injectionToken`. Not sure about it yet.

Sorry it took me so (incredibly) long to review. I didn't see this PR for years.

nicojs committed:
feat(child injector): Add `createChildInjector` (#72)

Add a new function to create a child injector, named `createChildInjector`. Contrary to its `provideXxx` counterparts,this will create a new disposable scope without providing additional injectable values.

nicojs committed:
chore(release): 5.0.0

yurahuchok commented:
> This is very interesting. I've got a couple of remarks.
> 
> 1. Could we also implement this for `provideFactory`?
> 2. Could we use overloading on the existing `provideClass` and `provideFactory` functions? I find the "provideInjectableXxx" naming scheme confusing, as classes provided by `provideClass` are also injectable.
> 3. Could we rename `injectableAs` to `knownAs`? Or maybe `injectionToken`. Not sure about it yet.
> 
> Sorry it took me so (incredibly) long to review. I didn't see this PR for years.

Thank you for the reply, and no worries on the delay.

1-2. Absolutely, I think the same thing can be implemented for "provideValue" and "provideFactory" without issue.

3. We can definitely rename "injectableAs" to "injectionToken". I agree it would provide additional clarity, since we already use the term "token", and this is essentially a wrapper for that.

I will update this pull request soon, and let you know.

_Off-topic: To be honest I generally found the implementation of "provideFactory" very confusing personally, and quite limiting when dealing with async bootstrapping (based on remote config values for example). Perhaps it can be updated to solve this issue, but I didn't spend any time on that. I instead use "factory" classes and inject them with "provideClass" and find it sufficient for the same use-case. But that's a whole different topic._

yurahuchok commented:
> This is very interesting. I've got a couple of remarks.
> 
> 1. Could we also implement this for `provideFactory`?
> 2. Could we use overloading on the existing `provideClass` and `provideFactory` functions? I find the "provideInjectableXxx" naming scheme confusing, as classes provided by `provideClass` are also injectable.
> 3. Could we rename `injectableAs` to `knownAs`? Or maybe `injectionToken`. Not sure about it yet.
> 
> Sorry it took me so (incredibly) long to review. I didn't see this PR for years.

Thank you for the reply, and no worries on the delay.

1. Absolutely, I think the same thing can be implemented for `provideFactory` without issue.

2. I agree on the topic of overloading too. I will investigate further, and implement if possible.

3. We can definitely rename `injectableAs` to `injectionToken`. I agree it would provide additional clarity, since we already use the term "token", and this is essentially a wrapper for that.

I will update this pull request soon, and let you know.

_Off-topic: Since you mentioned `provideFactory`, I feel compelled to voice my opinion here. To be honest I generally found the implementation of `provideFactory` very confusing personally, and quite limiting when dealing with async bootstrapping (based on remote config values for example). Perhaps it can be updated to solve this issue, but I didn't spend any time on that. I instead use "factory" classes and inject them with `provideClass` and find it sufficient for the same use-case. But that's a whole different topic._

nicojs commented:
> I will update this pull request soon, and let you know.

Thanks! I will keep an eye on this :)

> when dealing with async bootstrapping 

I've been thinking about `async` a lot. We can make an async variant of `provideFactory`, maybe called `provideAsync`? It must cascade all `injectXxx` methods since they must also be `async`. We might need to duplicate the code to make this work, but it is doable IMO

yurahuchok committed:
Merge branch 'master' of https://github.com/nicojs/typed-inject into nicojs-master-2

yurahuchok committed:
feat: provideClass overload to support internal knownAs tokens

yurahuchok committed:
chore: update README.md for Known Classes

yurahuchok committed:
feat: provideFactory overload to support internal knownAs tokens

yurahuchok committed:
chore: update README.md for Known Factories

yurahuchok commented:
Hi @nicojs !

How have you been? I finally got around to implementing the changes we talked about before. So sorry for the delay!

I have given it some thought and went with your original suggestion of calling this new property "knownAs", and subsequently calling this mechanics "Known Classes" and "Known Functions".

So as previously agreed:
1. Feature also implemented for `provideFactory`.
1. Overloading implemented for `provideClass` and `provideFactory`.
2. The internal token renamed to `knownAs`.

Please review this when you can, and let me know what you think. Particularly, about the branding of this feature.

Cheers!

yurahuchok committed:
fix: lint

yurahuchok committed:
fix: type check for provideClass and provideFactory overloading

yurahuchok committed:
chore: expand tests to include knownAs token

yurahuchok committed:
chore: improve test

yurahuchok committed:
chore: use better emoji in README.md for internally registered tokens

yurahuchok committed:
chore: update README.md

Xylenox commented on README.md:
```diff
@@ -405,6 +406,53 @@ class Foo {
+}
+fooFactory.knownAs = 'Foo' as const;
+
+function barFactory() {
```
This doesn't compile right? you need an input for the Foo dependency?
